### PR TITLE
Workflow for the ledger-api-bench-tool  defined in YAML file [DPP-669]

### PR DESCRIPTION
**Pull request by kamil-da**
### Motivation
We would like to be able to define submission and stream reading phases through a single YAML config file.

### Changes
- handling all parameters that were previously available only as CLI options in YAML workflow config file
- moved config-related classes to a new `config` package for clarity
- unified `WorkflowDescriptor` and `StreamConfig` classes into a single ADT tree with `WorkflowConfig` being the root

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/ConfigEnricher.scala [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+import com.daml.ledger.api.benchtool.config.WorkflowConfig.StreamConfig
+import com.daml.ledger.api.benchtool.submission.CommandSubmitter
+import com.daml.ledger.api.v1.value.Identifier
+import com.daml.ledger.test.model.Foo.{Foo1, Foo2, Foo3}
+import scalaz.syntax.tag._
+
+object ConfigEnricher {
+
+  def enrichedStreamConfig(
```
Comment (kamil-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/ConfigEnricher.scala:14]: This method enriches short party and template names with appropriate suffixes that are generated in the submission phase. For example, one can defined a submission phase using `Obs-1` as party. In reality a party name could be something like `Obs-1-1c33fa656d54e` with a generated suffix for uniqueness. `ConfigEnricher` will take care to pass correct party name.

Comment (kamil-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/ConfigEnricher.scala:14]: Please let me know if should put such comment in the code as well.

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/config/Cli.scala [hunk 2]
```diff
@@ -229,56 +234,33 @@ object Cli {
...
-            .map(templateIdFromString)
-            .foldLeft[Either[String, List[Identifier]]](Right(List.empty[Identifier])) {
-              case (acc, next) =>
-                for {
-                  ids <- acc
-                  id <- next
-                } yield id :: ids
-            }
-            .map(party -> _)
+          Right(WorkflowConfig.StreamConfig.PartyFilter(party, templates))
```
Comment (kamil-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/config/Cli.scala:260]: Template ids are now `String`s to allow flexible filter specification through the YAML file (one can define a template in the short form like: `Foo1`).

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/ConfigEnricher.scala (hunk 1 above)
Comment (mziolekda) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/ConfigEnricher.scala:14]: That's fine as is actually

Comment (mziolekda) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/ConfigEnricher.scala:14]: I would use imperative throughout the method names though
```suggestion
  def enrichStreamConfig(
```

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/config/Cli.scala [hunk 3]
```diff
@@ -109,7 +110,7 @@ object Cli {
     note(1, "Transactions/transaction trees:")
     note(2, "stream-type=<transactions|transaction-trees>", "(required)")
     note(2, "name=<stream-name>", "Stream name used to identify results (required)")
-    note(2, "filters=party1|template1|template2&party2", "(required)")
+    note(2, "filters=party1@template1@template2+party2", "(required)")
```
Comment (mziolekda) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/config/Cli.scala:113]: well spotted

**Review by mziolekda — APPROVED** [review]

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/ConfigEnricher.scala (hunk 1 above)
Comment (kamil-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/ConfigEnricher.scala]: Fixed 👍

**Pull request merged**